### PR TITLE
Migratable nodes option

### PR DESCRIPTION
**Pull request by chrisho**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**
https://github.com/harvester/harvester/issues/3032#issue-1422120838

**Solution:**
Filter migratable nodes through by VM network to choose from.

**Related Issue:**
https://github.com/harvester/harvester/issues/3032

**Test plan:**
1. Create a ClusterNetwork.
2. Create a Network Config(at least two nodes) for ClusterNetwork.
3. Create a VM Network with the ClusterNetwork.
4. Create a VM with the VM Network.
5. Get and choose a migratable node from API: https://harvester/v1/harvester/kubevirt.io.virtualmachines/default/{your_vm}?action=getMigratableNodes.
```
{
    "nodes": [
        "node2"
    ]
}
```
6. (Optional) verify failure, choose a node without the ClusterNetwork and migrate, and check the response for error.
image
7. Migrate to a migratable node and wait for the VM migration to succeed.

### CODE REVIEW

File: pkg/api/vm/handler.go [hunk 1]
```diff
@@ -41,6 +43,9 @@ const (
 	vmResource    = "virtualmachines"
 	vmiResource   = "virtualmachineinstances"
 	sshAnnotation = "harvesterhci.io/sshNames"
+
+	networkGroup      = "network.harvesterhci.io"
+	keyClusterNetwork = networkGroup + "/clusternetwork"
```
Comment (futuretea) [inline pkg/api/vm/handler.go]: can use https://github.com/harvester/network-controller-harvester/blob/7765e83555157622612e3a12f0df14204a3a97bf/pkg/utils/labels.go#L10 instead

File: pkg/api/vm/handler.go [hunk 2]
```diff
@@ -369,6 +382,24 @@ func (h *vmActionHandler) migrate(ctx context.Context, namespace, vmName string,
 	return err
 }
 
+func (h *vmActionHandler) isMigratableNode(targetNode string, vmi *kubevirtv1.VirtualMachineInstance) (bool, error) {
+	if targetNode == "" {
+		return true, nil
```
Comment (futuretea) [inline pkg/api/vm/handler.go:386]: if targetNode == “”， can we check if there has migratable nodes?

File: pkg/api/vm/handler.go [hunk 3]
```diff
@@ -397,6 +428,85 @@ func (h *vmActionHandler) abortMigration(namespace, name string) error {
...
+		Nodes []string `json:"nodes"`
+	}{
+		Nodes: nodes,
+	}
+
+	b, err := json.Marshal(resp)
+	if err != nil {
+		return err
+	}
+	_, err = rw.Write(b)
```
Comment (futuretea) [inline pkg/api/vm/handler.go]: should we return `Content-type` `application/json`?

File: pkg/api/vm/handler.go [hunk 4]
```diff
@@ -397,6 +428,85 @@ func (h *vmActionHandler) abortMigration(namespace, name string) error {
...
+		if requirement != nil {
+			nodeSelector = nodeSelector.Add(*requirement)
+		}
+	}
+	nodes, err := h.nodeCache.List(nodeSelector)
+	if err != nil {
+		return migratableNodes, err
+	}
+
+	// ignore the node where the VM is running
```
Comment (futuretea) [inline pkg/api/vm/handler.go:462]: var migratableNodes = make([]string, 0, len(nodes)-1)

File: pkg/api/vm/handler.go [hunk 5]
```diff
@@ -369,6 +382,24 @@ func (h *vmActionHandler) migrate(ctx context.Context, namespace, vmName string,
...
+	if targetNode == "" {
+		return true, nil
+	}
+
+	nodes, err := h.getMigratableNodesByVMI(vmi)
+	if err != nil {
+		return false, err
+	}
+
+	for _, node := range nodes {
```
Comment (futuretea) [inline pkg/api/vm/handler.go]: return slice.ContainsString(nodes, targetNode), nil

File: pkg/api/vm/handler.go [hunk 6]
```diff
@@ -397,6 +428,85 @@ func (h *vmActionHandler) abortMigration(namespace, name string) error {
 	return nil
 }
 
+func (h *vmActionHandler) getMigratableNodes(rw http.ResponseWriter, namespace, name string) error {
+	vmi, err := h.vmiCache.Get(namespace, name)
+	if err != nil {
+		return err
+	}
+	nodes, err := h.getMigratableNodesByVMI(vmi)
+	resp := struct {
```
Comment (futuretea) [inline pkg/api/vm/handler.go]: Prefer to add the Output struct to types.go

File: pkg/api/vm/schema.go [hunk 7]
```diff
@@ -50,6 +50,7 @@ func RegisterSchema(scaled *config.Scaled, server *server.Server, options config
 	vmtv := scaled.HarvesterFactory.Harvesterhci().V1beta1().VirtualMachineTemplateVersion()
 	vmImages := scaled.HarvesterFactory.Harvesterhci().V1beta1().VirtualMachineImage()
 	storageClasses := scaled.StorageFactory.Storage().V1().StorageClass()
+	nadCache := scaled.CniFactory.K8s().V1().NetworkAttachmentDefinition().Cache()
```
Comment (futuretea) [inline pkg/api/vm/schema.go]: nads := scaled.CniFactory.K8s().V1().NetworkAttachmentDefinition()

File: pkg/api/vm/schema.go [hunk 8]
```diff
@@ -77,6 +78,7 @@ func RegisterSchema(scaled *config.Scaled, server *server.Server, options config
 		backupCache:               backups.Cache(),
 		restores:                  restores,
 		settingCache:              settings.Cache(),
+		nadCache:                  nadCache,
```
Comment (futuretea) [inline pkg/api/vm/schema.go]: nadCache: nads.Cache(),

File: pkg/api/vm/handler.go [hunk 9]
```diff
@@ -331,6 +340,13 @@ func (h *vmActionHandler) migrate(ctx context.Context, namespace, vmName string,
 			return fmt.Errorf("Volume %s is container disk, needs to be removed before migration", volume.Name)
 		}
 	}
+
+	if ok, err := h.isMigratableNode(nodeName, vmi); !ok {
+		return errors.New("The target node is non-migratable")
```
Comment (futuretea) [inline pkg/api/vm/handler.go]: missing the err content

File: pkg/api/vm/formatter.go [hunk 10]
```diff
@@ -16,21 +16,22 @@ import (
...
+	startVM            = "start"
+	stopVM             = "stop"
+	restartVM          = "restart"
+	softReboot         = "softreboot"
+	pauseVM            = "pause"
+	unpauseVM          = "unpause"
+	ejectCdRom         = "ejectCdRom"
+	migrate            = "migrate"
+	abortMigration     = "abortMigration"
+	getMigratableNodes = "getMigratableNodes"
```
Comment (futuretea) [inline pkg/api/vm/formatter.go]: the http method of actions is `POST`, but the name `getMigratableNodes` seems like a GET API.
maybe `findMigratableNodes` would be better？

File: pkg/api/vm/handler.go [hunk 11]
```diff
@@ -397,6 +429,82 @@ func (h *vmActionHandler) abortMigration(namespace, name string) error {
 	return nil
 }
 
+func (h *vmActionHandler) findMigratableNodes(rw http.ResponseWriter, namespace, name string) error {
+	vmi, err := h.vmiCache.Get(namespace, name)
+	if err != nil {
+		return err
+	}
+	nodes, err := h.findMigratableNodesByVMI(vmi)
```
Comment (futuretea) [inline pkg/api/vm/handler.go:439]: this err missing check

**Review by futuretea — APPROVED** [review]
LGTM, Thanks! test plan passed

File: pkg/api/vm/handler.go [hunk 12]
```diff
@@ -331,6 +340,15 @@ func (h *vmActionHandler) migrate(ctx context.Context, namespace, vmName string,
 			return fmt.Errorf("Volume %s is container disk, needs to be removed before migration", volume.Name)
 		}
 	}
+
+	if ok, err := h.isMigratableNode(nodeName, vmi); nodeName == "" && err != nil {
```
Comment (guangbochen) [inline pkg/api/vm/handler.go]: the nodeName and err check should be separated; it is possible that nodeName is an empty string and err is not nil, but in that case, the error message returned would not indicate the actual problem, let's move it to the top method instead. Additionally, simply the check conditions into:

An updated pseudo code:
```
if ok, err := h.isMigratableNode(nodeName, vmi); err != nil {
	return fmt.Errorf("can't migrate the VM to the node %s: %s", nodeName, err.Error())
} else if !ok {
	return errors.New("the target node is non-migratable")
}
```

File: pkg/api/vm/handler.go [hunk 13]
```diff
@@ -369,6 +384,23 @@ func (h *vmActionHandler) migrate(ctx context.Context, namespace, vmName string,
...
 }
 
+func (h *vmActionHandler) isMigratableNode(targetNode string, vmi *kubevirtv1.VirtualMachineInstance) (bool, error) {
+	nodes, err := h.findMigratableNodesByVMI(vmi)
+	if err != nil {
+		return false, err
+	}
+
+	if len(nodes) == 0 {
+		return false, errors.New("no migration nodes")
```
Comment (guangbochen) [inline pkg/api/vm/handler.go]:
```suggestion
		return false, errors.New("no matching migratable nodes found")
```

File: pkg/api/vm/handler.go [hunk 14]
```diff
@@ -369,6 +384,23 @@ func (h *vmActionHandler) migrate(ctx context.Context, namespace, vmName string,
...
+	}
+
+	if targetNode == "" {
+		return true, nil
+	}
```
Comment (guangbochen) [inline pkg/api/vm/handler.go]: ditto, remove to the top

File: pkg/api/vm/handler_test.go [hunk 15]
```diff
@@ -161,14 +163,34 @@ func TestMigrateAction(t *testing.T) {
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "fake-node1",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "fake-node2",
+			},
+		},
+	}
```
Comment (guangbochen) [inline pkg/api/vm/handler_test.go:171]: can we add proper testing cases to validate the `findMigratableNodes` output?

File: pkg/api/vm/handler.go [hunk 16]
```diff
@@ -397,6 +429,85 @@ func (h *vmActionHandler) abortMigration(namespace, name string) error {
...
+			nodeSelector = nodeSelector.Add(*requirement)
+		}
+	}
+	nodes, err := h.nodeCache.List(nodeSelector)
+	if err != nil {
+		return nil, err
+	}
+	if len(nodes) == 0 {
+		return nil, nil
+	}
```
Comment (guangbochen) [inline pkg/api/vm/handler.go]: prefer to define the `migratableNodes` var on the top and return it here instead.

File: pkg/api/vm/handler.go [hunk 17]
```diff
@@ -397,6 +429,85 @@ func (h *vmActionHandler) abortMigration(namespace, name string) error {
...
+	for _, network := range vmNetworks {
+		requirement, err := h.getNodeSelectorRequirementFromNetwork(vmi.Namespace, network)
+		if err != nil {
+			return nil, err
+		}
+		if requirement != nil {
+			nodeSelector = nodeSelector.Add(*requirement)
+		}
+	}
+	nodes, err := h.nodeCache.List(nodeSelector)
```
Comment (guangbochen) [inline pkg/api/vm/handler.go]: will need to filter out nodes with not-ready, drained, and under maintenance conditions.

File: pkg/api/vm/handler.go [hunk 18]
```diff
@@ -397,6 +429,85 @@ func (h *vmActionHandler) abortMigration(namespace, name string) error {
...
+	for _, node := range nodes {
+		if vmi.Status.NodeName == node.Name {
+			continue
+		}
+		migratableNodes = append(migratableNodes, node.Name)
+	}
+	return migratableNodes, nil
+}
+
+func (h *vmActionHandler) getNodeSelectorRequirementFromNetwork(defaultNamespace string, net kubevirtv1.Network) (*labels.Requirement, error) {
```
Comment (guangbochen) [inline pkg/api/vm/handler.go]: You should be able to use the existing rules from the VM/VMI; you also need to add the user's custom selector rules into account.

**Comment by chrisho** [comment]
The latest commit has significant changes and needs to be reviewed again, thanks.

File: pkg/api/vm/formatter.go [hunk 19]
```diff
@@ -226,6 +228,13 @@ func canAbortMigrate(vmi *kubevirtv1.VirtualMachineInstance) bool {
 	return false
 }
 
+func isSpecificNodeFromVMINodeSelector(vmi *kubevirtv1.VirtualMachineInstance) bool {
```
Comment (guangbochen) [inline pkg/api/vm/formatter.go]: \s\isSpecificNodeFromVMINodeSelector\hasNodeSelector

File: pkg/api/vm/handler.go [hunk 20]
```diff
@@ -397,6 +427,93 @@ func (h *vmActionHandler) abortMigration(namespace, name string) error {
...
+
+		migratableNodes = append(migratableNodes, node.Name)
+	}
+	return migratableNodes, nil
+}
+
+func isDrained(node *corev1.Node) bool {
+	if node.Spec.Unschedulable {
+		return true
+	}
```
Comment (guangbochen) [inline pkg/api/vm/handler.go:487]: you may also check if the node already contains the [harvester-drained annotation](https://github.com/harvester/harvester/blob/master/pkg/util/drainhelper/helper.go#L31) and return it directly

File: pkg/api/vm/handler_test.go [hunk 21]
```diff
@@ -622,3 +644,239 @@ func (c fakeVirtualMachineInstanceMigrationClient) Watch(namespace string, opts
...
+	return result, err
+}
+
+func (c fakeNodeCache) AddIndexer(indexName string, indexer ctlcorev1.NodeIndexer) {
+	panic("implement me")
+}
+
+func (c fakeNodeCache) GetByIndex(indexName, key string) ([]*corev1.Node, error) {
+	panic("implement me")
+}
```
Comment (guangbochen) [inline pkg/api/vm/handler_test.go]: resue the fakeclient on the https://github.com/harvester/harvester/tree/master/pkg/util/fakeclients

File: pkg/api/vm/handler.go [hunk 22]
```diff
@@ -320,6 +325,10 @@ func (h *vmActionHandler) migrate(ctx context.Context, namespace, vmName string,
 	if !canMigrate(vmi) {
 		return errors.New("The VM is already in migrating state")
 	}
+	if isSpecificNodeFromVMINodeSelector(vmi) {
```
Comment (futuretea) [inline pkg/api/vm/handler.go]: check node selector in canMigrate.

---
for the vmi has node selector case,
UI will call findMigratableNodes and report an error when users click the `Migrate` button.
image
if we check node selector in `canMigrate`, the `Migrate` button will be hidden

File: pkg/api/vm/handler.go [hunk 23]
```diff
@@ -397,6 +427,93 @@ func (h *vmActionHandler) abortMigration(namespace, name string) error {
...
 }
 
+func (h *vmActionHandler) findMigratableNodes(rw http.ResponseWriter, namespace, name string) error {
+	vmi, err := h.vmiCache.Get(namespace, name)
+	if err != nil {
+		return err
+	}
+
+	if !canMigrate(vmi) {
+		return errors.New("The VM is already in migrating state")
```
Comment (futuretea) [inline pkg/api/vm/handler.go]: The error message does not match the function name

File: pkg/api/vm/types.go [hunk 24]
```diff
@@ -42,3 +42,7 @@ type RemoveVolumeInput struct {
 type CloneInput struct {
 	TargetVM string `json:"targetVm"`
 }
+
+type GetMigratableNodesOutput struct {
```
Comment (futuretea) [inline pkg/api/vm/types.go]: FindMigratableNodesOutput

**Review by futuretea — APPROVED** [review]
LGTM, Thanks. We need to figure out why CI is failing

**Review by guangbochen — DISMISSED** [review]
LGTM, thanks.

File: pkg/api/vm/handler.go [hunk 25]
```diff
@@ -397,6 +426,97 @@ func (h *vmActionHandler) abortMigration(namespace, name string) error {
...
+		}
+	}
+
+	return false
+}
+
+func (h *vmActionHandler) getNodeSelectorRequirementFromVMI(vmi *kubevirtv1.VirtualMachineInstance) (labels.Selector, error) {
+	if vmi == nil {
+		return nil, nil
+	}
```
Comment (futuretea) [inline pkg/api/vm/handler.go]:
```suggestion
	if vmi == nil || vmi.Spec.Affinity == nil || vmi.Spec.Affinity.NodeAffinity == nil || vmi.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
		return labels.Everything(), nil
	}
```

File: pkg/api/vm/formatter.go [hunk 26]
```diff
@@ -206,7 +208,8 @@ func canMigrate(vmi *kubevirtv1.VirtualMachineInstance) bool {
 		vmi.Annotations[util.AnnotationMigrationUID] == "" {
 		return true
 	}
-	return false
+
+	return vmi.Spec.NodeSelector == nil
```
Comment (futuretea) [inline pkg/api/vm/formatter.go]:
```go
func canMigrate(vmi *kubevirtv1.VirtualMachineInstance) bool {
	if vmi == nil {
		return false
	}

	if !vmi.IsRunning() {
		return false
	}

	// The VM is already in migrating state
	if vmi.Annotations[util.AnnotationMigrationUID] != "" {
		return false
	}

	// The VM is set to run on a specific node
	if vmi.Spec.NodeSelector != nil && vmi.Spec.NodeSelector[corev1.LabelHostname] != "" {
		return false
	}

	return true
}
```

Comment (futuretea) [reply pkg/api/vm/formatter.go]: fix CI failed. due to `vmi.Spec.NodeSelector` panic when 
```
	if canMigrate(vmi) {
		resource.AddAction(request, migrate)
		resource.AddAction(request, findMigratableNodes)
	}
```

File: pkg/api/vm/handler.go (hunk 25 above)
Comment (futuretea) [reply pkg/api/vm/handler.go]: if the vm is using management Network, `vmi.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms` will panic
and we also need return `labels.Everything()` is this case. if we return `nil`, `nodeCache.List(nodeSelector)` will panic too.

**Review by futuretea — DISMISSED** [review]
drone CI failed

**Review by guangbochen — APPROVED** [review]

**Review by futuretea — APPROVED** [review]

**Pull request merged**